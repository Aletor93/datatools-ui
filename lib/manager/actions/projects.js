import {createAction} from 'redux-actions'
import {browserHistory} from 'react-router'

import {secureFetch} from '../../common/actions'
import {getConfigProperty} from '../../common/util/config'
import {fetchProjectFeeds} from './feeds'
import {updateGtfsFilter} from '../../gtfs/actions/filter'
import {handleJobResponse} from './status'
import {getActiveProject} from '../selectors'
import {setVisibilitySearchText} from './visibilityFilter'

// Bulk Project Actions
const receiveProjects = createAction('RECEIVE_PROJECTS')
const requestingProjects = createAction('REQUESTING_PROJECTS')
// Single Project Actions
const requestingProject = createAction('REQUESTING_PROJECT')
const deletingProject = createAction('DELETING_PROJECT')
const requestingSync = createAction('REQUESTING_SYNC')
const receiveSync = createAction('RECEIVE_SYNC')
const savingProject = createAction('SAVING_PROJECT')
// Also used in deployments.js (actions)
export const receiveProject = createAction('RECEIVE_PROJECT')
const settingActiveProject = createAction('SET_ACTIVE_PROJECT')
const deletedProject = createAction('DELETED_PROJECT')
const runningFetchFeedsForProject = createAction('RUNNING_FETCH_FEED_FOR_PROJECT')
// Used by component
export const createProject = createAction('CREATE_PROJECT')

export function setActiveProject (project) {
  return function (dispatch, getState) {
    return dispatch(fetchProjectFeeds(project.id))
      .then(() => {
        dispatch(settingActiveProject(project))
        return dispatch(updateGtfsFilter(getActiveProject(getState()), getState().user))
      })
  }
}

export function fetchProjects (getActive = false) {
  return function (dispatch, getState) {
    dispatch(requestingProjects())
    return dispatch(secureFetch('/api/manager/secure/project'))
      .then(response => response.json())
      .then(projects => {
        dispatch(receiveProjects(projects))
        // return active project if requested
        if (getActive) {
          const activeProject = getActiveProject(getState())
          if (!activeProject.feedSources) {
            return dispatch(fetchProjectFeeds(activeProject.id))
              .then(() => {
                return dispatch(updateGtfsFilter(getActiveProject(getState()), getState().user))
                  .then(() => {
                    return activeProject
                  })
              })
          }
        }
        return projects
      })
  }
}

export function fetchProjectsWithPublicFeeds () {
  return function (dispatch, getState) {
    dispatch(requestingProjects())
    const url = '/api/manager/public/project'
    return dispatch(secureFetch(url))
      .then(response => response.json())
      .then(projects => dispatch(receiveProjects(projects)))
  }
}

export function fetchProject (projectId, unsecure) {
  return function (dispatch, getState) {
    dispatch(requestingProject({projectId}))
    const apiRoot = unsecure ? 'public' : 'secure'
    const url = `/api/manager/${apiRoot}/project/${projectId}`
    return dispatch(secureFetch(url))
      .then(response => response.json())
      .then(project => {
        dispatch(receiveProject(project))
        return project
      })
  }
}

export function fetchProjectWithFeeds (projectId, unsecure) {
  return function (dispatch, getState) {
    dispatch(requestingProject({projectId}))
    const apiRoot = unsecure ? 'public' : 'secure'
    const url = `/api/manager/${apiRoot}/project/${projectId}`
    return dispatch(secureFetch(url))
      .then(response => response.json())
      .then(project => {
        dispatch(receiveProject(project))
        if (!unsecure) {
          return dispatch(fetchProjectFeeds(project.id))
        }
      })
  }
}

export function deleteProject (project) {
  return function (dispatch, getState) {
    dispatch(deletingProject(project))
    const url = `/api/manager/secure/project/${project.id}`
    return dispatch(secureFetch(url, 'delete'))
      .then(response => response.json())
      .then(project => dispatch(deletedProject(project)))
      .then(() => browserHistory.push(`/home`))
  }
}

export function updateProject (project, changes, fetchFeeds = false) {
  return function (dispatch, getState) {
    dispatch(savingProject())
    const url = `/api/manager/secure/project/${project.id}`
    return dispatch(secureFetch(url, 'put', changes))
      .then((res) => {
        if (fetchFeeds) {
          return dispatch(fetchProjectWithFeeds(project.id))
        } else {
          return dispatch(fetchProject(project.id))
        }
      })
  }
}

export function deployPublic (project) {
  return function (dispatch, getState) {
    // dispatch(savingProject())
    const url = `/api/manager/secure/project/${project.id}/deployPublic`
    return dispatch(secureFetch(url, 'post'))
      .then((res) => res.json())
      .then(json => {
        return json
      })
  }
}

export function thirdPartySync (projectId, type) {
  return function (dispatch, getState) {
    dispatch(requestingSync(projectId))
    const url = `/api/manager/secure/project/${projectId}/thirdPartySync/${type}`
    return dispatch(secureFetch(url))
      .then(response => response.json())
      // .catch(err => console.log(err))
      .then(project => {
        dispatch(receiveSync(projectId))
        return dispatch(fetchProjectWithFeeds(projectId))
      })
  }
}

export function fetchFeedsForProject (project) {
  return function (dispatch, getState) {
    dispatch(runningFetchFeedsForProject(project))
    const url = `/api/manager/secure/project/${project.id}/fetch`
    return dispatch(secureFetch(url, 'post'))
      .then(res => dispatch(handleJobResponse(res, 'Error fetching project feeds')))
  }
}

export function saveProject (props) {
  return function (dispatch, getState) {
    dispatch(savingProject(props))
    const url = '/api/manager/secure/project'
    return dispatch(secureFetch(url, 'post', props))
      .then((res) => res.json())
      .then((json) => dispatch(fetchProjects()))
  }
}

/**
 * Download a merged GTFS file for a Project
 */
export function downloadFeedForProject (project) {
  return function (dispatch, getState) {
    const url = `/api/manager/secure/project/${project.id}/download`
    // window.location.assign(url)
    return dispatch(secureFetch(url))
      .then(res => dispatch(handleJobResponse(res, 'Error merging project feeds')))
  }
}

/**
 * Download a GTFS file for a merged project feed.
 */
export function downloadMergedFeedViaToken (project, isPublic, prefix) {
  return function (dispatch, getState) {
    const route = isPublic ? 'public' : 'secure'
    const url = `/api/manager/${route}/project/${project.id}/downloadtoken`
    return dispatch(secureFetch(url))
      .then(response => response.json())
      .then(json => {
        if (getConfigProperty('application.data.use_s3_storage')) {
          // Download object using presigned S3 URL.
          window.location.assign(json.url)
        } else {
          // use token to download feed
          window.location.assign(`/api/manager/downloadprojectfeed/${json.id}`)
        }
      })
  }
}

/**
 * When project viewer component mounts, unconditionally load project with feed
 * sources to ensure that updates to feed sources (e.g., a new feed version) are
 * applied to other project collections (e.g., a deployment that references said
 * feed source).
 */
export function onProjectViewerMount (initialProps) {
  return function (dispatch, getState) {
    const {projectId} = initialProps.router.params
    dispatch(setVisibilitySearchText(null))
    dispatch(fetchProjectWithFeeds(projectId))
  }
}
