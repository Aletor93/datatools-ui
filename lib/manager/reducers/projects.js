import update from 'react-addons-update'
import {getConfigProperty} from '../../common/util/config'
import {defaultSorter} from '../../common/util/util'

const defaultState = {
  isFetching: false,
  all: null,
  active: null,
  filter: {
    searchText: null
  }
}

// FIXME Use handleActions to simplify reducer syntax
const projects = (state = defaultState, action) => {
  // let projects, projectIndex, sourceIndex, versionIndex, activeProject, activeIndex, feeds, activeProjectId, project, newState, deployment
  switch (action.type) {
    case 'SET_PROJECT_VISIBILITY_SEARCH_TEXT':
      return update(state, {filter: {searchText: {$set: action.text}}})
    case 'SET_PROJECT_VISIBILITY_FILTER':
      return update(state, {filter: {filter: {$set: action.filter}}})
    case 'CREATE_PROJECT': {
      const projects = [{
        isCreating: true,
        name: '',
        ...action.payload
      }, ...state.all]
      return update(state, { all: { $set: projects } })
    }
    case 'CREATE_FEEDSOURCE': {
      const {projectIndex} = state.all.findIndex(p => p.id === action.projectId)
      const project = state.all[projectIndex]
      let newState = null
      // console.log('adding fs to project', state.all[projectIndex])

      // if project's feedSources array is undefined, add it
      if (!project.feedSources) {
        // console.log('adding new fs array')
        newState = update(state, {all: {[projectIndex]: {$merge: {feedSources: []}}}})
      }

      // add new empty feed source to feedSources array
      const feedSource = {
        isCreating: true,
        name: '',
        projectId: project.id
      }
      return update(newState || state, {
        all: {[projectIndex]: {feedSources: {$unshift: [feedSource]}}}
      })
    }
    case 'CREATE_DEPLOYMENT': {
      const projectIndex = state.all.findIndex(p => p.id === action.projectId)
      const project = state.all[projectIndex]
      let newState
      // if project's deployment array is undefined, add it
      if (!project.deployments) {
        newState = update(state, {
          all: {[projectIndex]: {$merge: {deployments: []}}}
        })
      }
      // add new empty feed source to feedSources array
      const deployment = {
        isCreating: true,
        name: '',
        project: project
      }
      return update(newState || state, {
        all: {[projectIndex]: {deployments: {$unshift: [deployment]}}}
      })
    }
    case 'REQUESTING_FEEDSOURCE':
    case 'REQUESTING_FEEDSOURCES':
    case 'REQUEST_PROJECTS': {
      return update(state, { isFetching: { $set: true } })
    }
    case 'RECEIVE_PROJECTS': {
      const activeProjectId = state.active
        ? state.active.id
        : getConfigProperty('application.active_project')
      let activeIndex = action.payload.findIndex(p => p.id === activeProjectId)
      if (activeIndex === -1) {
        activeIndex = 0
      }
      return {
        isFetching: false,
        all: action.payload,
        active: action.payload[activeIndex] && action.payload[activeIndex].id,
        filter: {
          searchText: null
        }
      }
    }
    case 'SET_ACTIVE_PROJECT': {
      return update(state, {
        active: { $set: action.payload && action.payload.id }
      })
    }
    // case 'DELETED_PROJECT':
    //   projectIndex = state.all.findIndex(p => p.id === action.payload.id)
    //   return update(state, {
    //     active: { $set: activeProject }, all: { splice: projects }
    //   })
    case 'RECEIVE_PROJECT': {
      let projects, projectIndex
      if (!action.payload) {
        return state
      }
      if (!state.all) {
        // there are no current projects loaded
        projects = [action.payload]
      } else {
        // projects already loaded
        projectIndex = state.all.findIndex(p => p.id === action.payload.id)
        if (projectIndex === -1) {
          // projects loaded but not this one; add it
          projects = [
            ...state.all,
            action.payload
          ]
        } else {
          // FIXME: Use $splice
          // projects loaded including this one, replace it
          projects = [
            ...state.all.slice(0, projectIndex),
            action.payload,
            ...state.all.slice(projectIndex + 1)
          ]
        }
      }
      const activeProjectId = state.active
        ? state.active.id
        : getConfigProperty('application.active_project')
      let activeProject = action.project.id === activeProjectId
        ? action.payload
        : null
      if (state.active && !activeProject) {
        // Active project already exists and received project does not match
        // active project.
        activeProject = state.active
      }
      return update(state, {active: { $set: activeProject }, all: { $set: projects }})
    }
    case 'RECEIVE_FEEDSOURCES': {
      const projectIndex = state.all.findIndex(p => p.id === action.projectId)
      let sourceIndex, feeds
      // If lazy fetching the rest of the feed sources, don't overwrite current
      // feed source.
      if (
        state.all[projectIndex] &&
        state.all[projectIndex].feedSources &&
        state.all[projectIndex].feedSources.length === 1
      ) {
        sourceIndex = action.feedSources
          .findIndex(fs => fs.id === state.all[projectIndex].feedSources[0].id)
        action.feedSources.splice(sourceIndex, 1)
        feeds = [
          ...state.all[projectIndex].feedSources,
          ...action.feedSources
        ]
      } else {
        feeds = action.feedSources
      }
      feeds = feeds.sort(defaultSorter)
      if (state.active && action.projectId === state.active.id) {
        return update(state, {
          isFetching: {$set: false},
          active: {$merge: {feedSources: feeds}},
          all: {
            [projectIndex]: {$merge: {feedSources: feeds}}
          }
        })
      } else if (feeds.length) {
        // If projectId does not match active project
        return update(state, {
          isFetching: {$set: false},
          all: {
            [projectIndex]: {$merge: {feedSources: feeds}}
          }
        })
      } else {
        return update(state, {isFetching: {$set: false}})
      }
    }
    case 'RECEIVE_FEEDSOURCE': {
      if (!action.feedSource) {
        return update(state, {isFetching: {$set: false}})
      }
      const projectIndex = state.all.findIndex(p => p.id === action.feedSource.projectId)
      const existingSources = state.all[projectIndex].feedSources || []
      const sourceIndex = existingSources.findIndex(s => s.id === action.feedSource.id)
      const updatedSources = sourceIndex !== -1
        ? {[sourceIndex]: {$set: action.feedSource}}
        : {$set: [action.feedSource]}
      return update(state, {
        all: {
          [projectIndex]: {
            feedSources: updatedSources
          }
        },
        isFetching: { $set: false }
      })
    }
    case 'RECEIVE_FEEDVERSIONS': {
      const projectIndex = state.all.findIndex(p => p.id === action.feedSource.projectId)
      const sourceIndex = state.all[projectIndex].feedSources.findIndex(s => s.id === action.feedSource.id)
      const versionSort = (a, b) => {
        if (a.version < b.version) return -1
        if (a.version > b.version) return 1
        return 0
      }
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {$merge: {
          feedVersions: action.feedVersions.sort(versionSort)
        }}}}}
      })
    }
    case 'RECEIVE_FEEDVERSION': {
      const projectIndex = state.all.findIndex(p => p.id === action.feedVersion.feedSource.projectId)
      const sourceIndex = state.all[projectIndex].feedSources.findIndex(s => s.id === action.feedVersion.feedSource.id)
      // console.log(sourceIndex)
      const versionIndex = state.all[projectIndex].feedSources[sourceIndex].feedVersions
        ? state.all[projectIndex].feedSources[sourceIndex].feedVersions.findIndex(v => v.id === action.feedVersion.id)
        : -1
      const updatedVersions = versionIndex !== -1
        ? {[versionIndex]: {$set: action.feedVersion}}
        : {$set: [action.feedVersion]}
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {
          feedVersions: updatedVersions
        }}}}
      })
    }
    case 'PUBLISHED_FEEDVERSION': {
      const projectIndex = state.all.findIndex(p => p.id === action.feedVersion.feedSource.projectId)
      const sourceIndex = state.all[projectIndex].feedSources.findIndex(s => s.id === action.feedVersion.feedSource.id)
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {
          publishedVersionId: {$set: action.feedVersion.id}
        }}}}
      })
    }
    case 'RECEIVE_VALIDATION_RESULT': {
      const projectIndex = state.all.findIndex(p => p.id === action.payload.feedVersion.feedSource.projectId)
      const sourceIndex = state.all[projectIndex].feedSources.findIndex(s => s.id === action.payload.feedVersion.feedSource.id)
      const versionIndex = state.all[projectIndex].feedSources[sourceIndex].feedVersions.findIndex(v => v.id === action.payload.feedVersion.id)
      // let result = {}
      // action.validationResult.map(error => {
      //   if (!result[error.file]) {
      //     result[error.file] = []
      //   }
      //   result[error.file].push(error)
      // })
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {feedVersions: {
          [versionIndex]: {$merge: {validationResult: action.payload.validationResult}
          }}}}}}
      })
    }
    case 'RECEIVE_VALIDATION_ERRORS': {
      const projectIndex = state.all.findIndex(p => p.id === action.payload.feedVersion.feedSource.projectId)
      const sourceIndex = state.all[projectIndex].feedSources.findIndex(s => s.id === action.payload.feedVersion.feedSource.id)
      const versionIndex = state.all[projectIndex].feedSources[sourceIndex].feedVersions.findIndex(v => v.id === action.payload.feedVersion.id)
      const errorIndex = state.all[projectIndex].feedSources[sourceIndex].feedVersions[versionIndex].validationResult.error_counts.findIndex(e => e.type === action.payload.errorType)
      const newErrors = state.all[projectIndex].feedSources[sourceIndex].feedVersions[versionIndex].validationResult.error_counts[errorIndex].errors
        ? {$push: action.payload.errors}
        : {$set: action.payload.errors}
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {feedVersions: {
          [versionIndex]: {validationResult: {error_counts: {
            [errorIndex]: {errors: newErrors}
          }}}}}}}
        }
      })
    }
    case 'RECEIVE_FEEDVERSION_ISOCHRONES': {
      const projectIndex = state.all.findIndex(p => p.id === action.feedSource.projectId)
      const sourceIndex = state.all[projectIndex].feedSources.findIndex(s => s.id === action.feedSource.id)
      const versionIndex = state.all[projectIndex].feedSources[sourceIndex].feedVersions.findIndex(v => v.id === action.feedVersion.id)
      const { fromLat, fromLon, date, fromTime, toTime } = action
      action.isochrones.properties = { fromLat, fromLon, date, fromTime, toTime }
      action.isochrones.features = action.isochrones && action.isochrones.features
        ? action.isochrones.features.map(f => {
          f.type = 'Feature'
          return f
        })
        : null
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {feedVersions: {
          [versionIndex]: {$merge: {isochrones: action.isochrones
          }}}}}}}
      })
    }
    case 'RECEIVE_DEPLOYMENTS': {
      const projectIndex = state.all.findIndex(p => p.id === action.projectId)
      if (state.active && action.projectId === state.active.id) {
        return update(state, {
          active: {$merge: {deployments: action.deployments}},
          all: {[projectIndex]: {$merge: {deployments: action.deployments}}}
        })
      } else {
        // if projectId does not match active project
        return update(state, {
          all: {[projectIndex]: {$merge: {deployments: action.deployments}}}
        })
      }
    }
    case 'RECEIVE_DEPLOYMENT': {
      const projectIndex = state.all
        .findIndex(p => p.id === action.deployment.project.id)
      const existingDeployments = state.all[projectIndex].deployments || []
      let updatedDeployments
      const sourceIndex = existingDeployments
        .findIndex(s => s.id === action.deployment.id)
      if (sourceIndex === -1) { // source does not currently; add it
        updatedDeployments = [
          ...existingDeployments,
          action.deployment
        ]
      } else { // existing feedsource array includes this one, replace it
        updatedDeployments = [
          ...existingDeployments.slice(0, sourceIndex),
          action.deployment,
          ...existingDeployments.slice(sourceIndex + 1)
        ]
      }
      return update(state, {
        all: {[projectIndex]: {$merge: {deployments: updatedDeployments}}}
      })
    }
    case 'RECEIVE_NOTES_FOR_FEEDSOURCE': {
      const projectIndex = state.all.findIndex(p => p.id === action.feedSource.projectId)
      const sourceIndex = state.all[projectIndex].feedSources.findIndex(s => s.id === action.feedSource.id)
      const {notes} = action
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {$merge: {notes}}}}}
      })
    }
    case 'RECEIVE_NOTES_FOR_FEEDVERSION': {
      const projectIndex = state.all.findIndex(p => p.id === action.feedVersion.feedSource.projectId)
      const sourceIndex = state.all[projectIndex].feedSources.findIndex(s => s.id === action.feedVersion.feedSource.id)
      const versionIndex = state.all[projectIndex].feedSources[sourceIndex].feedVersions.findIndex(v => v.id === action.feedVersion.id)
      return update(state, {all: {
        [projectIndex]: {feedSources: {[sourceIndex]: {feedVersions: {
          [versionIndex]: {$merge: {notes: action.notes}
          }}}}}}
      })
    }
    case 'RECEIVE_GTFSEDITOR_SNAPSHOTS': {
      const projectIndex = state.all.findIndex(p => p.id === action.payload.feedSource.projectId)
      const sourceIndex = state.all[projectIndex].feedSources.findIndex(s => s.id === action.payload.feedSource.id)
      return update(state, {
        all: {
          [projectIndex]: {
            feedSources: {
              [sourceIndex]: {
                $merge: {
                  editorSnapshots: action.payload.snapshots
                }
              }
            }
          }
        }
      })
    }
    default:
      return state
  }
}

// TODO: Use this function to get indexes (perhaps set to idx variable?)
// function getIndexesFromFeed ({state, feedVersion, feedSource, projectId}) {
//   if (!feedSource && feedVersion) ({feedSource} = feedVersion)
//   if (!projectId) ({projectId} = feedSource)
//   const projectIndex = state.all.findIndex(p => p.id === projectId)
//   const sources = state.all[projectIndex].feedSources || []
//   const sourceIndex = feedSource && sources.findIndex(s => s.id === feedSource.id)
//   const versionIndex = feedVersion && sources[sourceIndex].feedVersions
//     ? sources[sourceIndex].feedVersions.findIndex(v => v.id === feedVersion.id)
//     : -1
//   return {
//     projectIndex,
//     sourceIndex,
//     versionIndex
//   }
// }

export default projects
