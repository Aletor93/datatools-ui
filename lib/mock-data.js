// @flow

import multi from '@conveyal/woonerf/store/multi'
import promise from '@conveyal/woonerf/store/promise'
import {middleware as fetchMiddleware} from '@conveyal/woonerf/fetch'
import clone from 'lodash/cloneDeep'
import {browserHistory} from 'react-router'
import {routerMiddleware, routerReducer as routing} from 'react-router-redux'
import {applyMiddleware, combineReducers, createStore} from 'redux'
import thunkMiddleware from 'redux-thunk'

import admin from './admin/reducers'
import {defaultState as defaultAdminOrganizationsState} from './admin/reducers/organizations'
import {defaultState as defaultAdminUsersState} from './admin/reducers/users'
import alerts from './alerts/reducers'
import {defaultState as defaultAlertsActiveState} from './alerts/reducers/active'
import {defaultState as defaultAlertsAlertsState} from './alerts/reducers/alerts'
import editor from './editor/reducers'
import {defaultState as defaultEditorDataState} from './editor/reducers/data'
import {defaultState as defaultEditorMapState} from './editor/reducers/mapState'
import {defaultState as defaultEditorSettingsState} from './editor/reducers/settings'
import {defaultState as defaultEditorTimetableState} from './editor/reducers/timetable'
import gtfs from './gtfs/reducers'
import {defaultState as defaultGtfsFilterState} from './gtfs/reducers/filter'
import {defaultState as defaultGtfsPatternState} from './gtfs/reducers/patterns'
import {defaultState as defaultGtfsRoutesState} from './gtfs/reducers/routes'
import {defaultState as defaultGtfsShapesState} from './gtfs/reducers/shapes'
import {defaultState as defaultGtfsStopsState} from './gtfs/reducers/stops'
import {defaultState as defaultGtfsTimetablesState} from './gtfs/reducers/timetables'
import {defaultState as defaultGtfsValidationState} from './gtfs/reducers/validation'
import * as gtfsPlusReducers from './gtfsplus/reducers'
import {defaultState as defaultGtfsPlusState} from './gtfsplus/reducers/gtfsplus'
import * as managerReducers from './manager/reducers'
import {defaultState as defaultManagerLanguagesState} from './manager/reducers/languages'
import {defaultState as defaultManagerProjectsState} from './manager/reducers/projects'
import {defaultState as defaultManagerStatusState} from './manager/reducers/status'
import {defaultState as defaultManagerUiState} from './manager/reducers/ui'
import {defaultState as defaultManagerUserState} from './manager/reducers/user'
import signs from './signs/reducers'
import {defaultState as defaultSignActiveState} from './signs/reducers/active'
import {defaultState as defaultSignSignsState} from './signs/reducers/signs'
import type {AppState} from './types/reducers'

const defaultManagerState = {
  languages: defaultManagerLanguagesState,
  projects: defaultManagerProjectsState,
  status: defaultManagerStatusState,
  ui: defaultManagerUiState,
  user: defaultManagerUserState
}

/**
 * Returns a cloned version of the initial load state of the redux store.
 */
export function getMockInitialState(): AppState {
  return clone({
    admin: {
      organizations: defaultAdminOrganizationsState,
      users: defaultAdminUsersState
    },
    alerts: {
      ...defaultAlertsActiveState,
      ...defaultAlertsAlertsState
    },
    ...defaultManagerState,
    editor: {
      data: defaultEditorDataState,
      // The editSettings state below represents the default state of redux-undo
      // which this reducer uses.
      editSettings: {
        future: [],
        present: defaultEditorSettingsState,
        index: 0,
        limit: 1,
        past: []
      },
      mapState: defaultEditorMapState,
      timetable: defaultEditorTimetableState
    },
    gtfs: {
      filter: defaultGtfsFilterState,
      patterns: defaultGtfsPatternState,
      routes: defaultGtfsRoutesState,
      shapes: defaultGtfsShapesState,
      stops: defaultGtfsStopsState,
      timetables: defaultGtfsTimetablesState,
      validation: defaultGtfsValidationState
    },
    gtfsplus: defaultGtfsPlusState,
    routing: {}, // react-router state
    signs: {
      ...defaultSignActiveState,
      ...defaultSignSignsState
    }
  })
}

/**
 * Returns a cloned version of the state or the redux store with a feed loaded
 * for editing in the editor.
 */
export function getMockEditorState (): AppState {
  const editorState = getMockInitialState()
  return clone(editorState)
}

/**
 * Creates a new store that has reducers that will update state.
 */
export function makeMockStore (initialState: AppState = getMockInitialState()) {
  return createStore(
    // $FlowFixMe not sure why this is giving a flow error
    combineReducers({
      routing,
      ...{
        ...managerReducers,
        admin,
        alerts,
        signs,
        ...gtfsPlusReducers,
        editor,
        gtfs
      }
    }),
    initialState,
    applyMiddleware(...[
      fetchMiddleware,
      multi,
      promise,
      thunkMiddleware
    ])
  )
}
