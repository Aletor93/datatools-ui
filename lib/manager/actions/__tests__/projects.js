// @flow

import {getMockInitialState, makeMockStore} from '../../../mock-data'
import * as projectsActions from '../projects'

function makeMockManagerStore () {
  return makeMockStore(getMockInitialState())
}

describe('manager > actions > projects >', () => {
  it('should set a new feed source display type', () => {
    const store = makeMockManagerStore()
    store.dispatch(projectsActions.toggleFeedTableViewType())
    store.expectStateToMatchSnapshot('projects')
  })

  it('should set a new feed source sort order', () => {
    const store = makeMockManagerStore()
    store.dispatch(projectsActions.setFeedSort('alphabetically-desc'))
    store.expectStateToMatchSnapshot('projects')
  })
})
