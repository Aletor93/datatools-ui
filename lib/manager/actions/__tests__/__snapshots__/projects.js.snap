// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager > actions > projects > should set a new feed source display type 1`] = `
Object {
  "active": null,
  "all": Array [],
  "feedTableViewType": "detailed",
  "filter": Object {
    "filter": null,
    "searchText": null,
  },
  "isFetching": false,
  "sort": "alphabetically-asc",
}
`;

exports[`manager > actions > projects > should set a new feed source sort order 1`] = `
Object {
  "active": null,
  "all": Array [],
  "feedTableViewType": "compact",
  "filter": Object {
    "filter": null,
    "searchText": null,
  },
  "isFetching": false,
  "sort": "alphabetically-desc",
}
`;
